docs/closure-compiler.md

```diff
@@ -0,0 +1,697 @@
+  getOffsetWidth() {}
+}
+```
+
+This is the syntax we use for specifying [structural types](https://github.com/google/closure-compiler/wiki/Structural-Interfaces-in-Closure-Compiler) within closure. The class methods, their parameters, and corresponding JSDoc specify the shape of an
+object that must contain these methods with their specified parameters and return values. This is
+mostly used to specify the shape of adapters, as mentioned above.
+
+### `@typedef` via `let` declaration
+
+**Example**:
+
+```js
+/**
+ * @typedef {{
+ *   isActivated: boolean,
+ *   wasActivatedByPointer: boolean,
+ *   wasElementMadeActive: boolean,
+ *   activationStartTime: number,
+ *   activationEvent: ?Event
+ * }}
+ */
+let ActivationStateType;
+
+/**
+ * @typedef {{foo: number}}
+ */
+export let MyExportedType;
+```
+
+While these `let` declarations do not do anything at runtime, they are used by closure to
+encapsulate complex types as specified through a [\@typedef statement](https://github.com/google/closure-compiler/wiki/Types-in-the-Closure-Type-System#typedefs). The statements above let both `ActivationStateType` and `MyExportedType` be used as type
+parameters throughout the rest of the code.
+
+### "Dictionary" objects via quoted keys / bracket references
+
+**Example**:
+
+```js
+/** @dict */
+const SETTINGS = {
+  'numRetries': 1,
+  'selectorToQuery': 'body',
+  'windowObject': window,
+};
+
+window.settings = SETTINGS;
+
+/** @const {!Object<string, string>} **/
+const DEACTIVATION_ACTIVATION_PAIRS = {
+  'mouseup': 'mousedown',
+  'pointerup': 'pointerdown',
+  'touchend': 'touchstart',
+  'keyup': 'keydown',
+  'blur': 'focus',
+};
+
+Object.keys(DEACTIVATION_ACTIVATION_PAIRS).forEach((deactivationEvt) => {
+  const activationEvt = DEACTIVATION_ACTIVATION_PAIRS[deactivationEvt];
+  domNode.addEventListener(activationEvt, someActivationListener);
+  domNode.addEventListener(deactivationEvt, someDeactivationListener);
+});
+
+/** @const {!Array<!Object<string, !Function>>} */
+const listeners = [
+  {
+    'mouseup': () => console.log('mouseup'),
+    'mousedown': () => console.log('mousedown'),
+  },
+  {
+    'keyup': () => console.log('keyup'),
+    'keydown': () => console.log('keydown'),
+  },
+];
+
+console.log(listeners[0]['mouseup']);
+```
+
+Sometimes in our code, object keys will have meaning outside just being a key for an object. An
+example of this might be a map of event types to their respective listeners, as shown above.
+
+When closure compiles javascript using advanced optimizations, it [obfuscates property values](https://developers.google.com/closure/compiler/docs/api-tutorial3#better) in
+objects, classes, etc. While this leads to smaller code, it also leads to issues when those
+properties are used externally. In order to prevent this behavior, object keys need to be quoted,
+and those keys need to be referenced using bracket notation.
+
+### Member declarations via sentinel property reference expressions.
+
+**Example:**
```

Comment:
I tried to keep all the examples at the top for consistency, but I see your point here. Moving it.